##I make a little experiment for this dataset and make a neural network classifying using scikit-learn library.



In [1]:
import pandas as pd


In [2]:
df = pd.read_csv('/content/AirQualityUCI.csv')


In [3]:
df.head()


Date      Time  CO(GT)  PT08.S1(CO)  ...  PT08.S5(O3)     T    RH      AH
0  10/3/2004  18.00.00     2.6         1360  ...       1268.0  13.6  48.9  0.7578
1  10/3/2004  19.00.00     2.0         1292  ...        972.0  13.3  47.7  0.7255
2  10/3/2004  20.00.00     2.2         1402  ...       1074.0  11.9  54.0  0.7502
3  10/3/2004  21.00.00     2.2         1376  ...       1203.0  11.0  60.0  0.7867
4  10/3/2004  22.00.00     1.6         1272  ...       1110.0  11.2  59.6  0.7888

[5 rows x 15 columns]

In [4]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   int64  
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   int64  
 7   NOx(GT)        9357 non-null   int64  
 8   PT08.S3(NOx)   9357 non-null   int64  
 9   NO2(GT)        9357 non-null   int64  
 10  PT08.S4(NO2)   9357 non-null   int64  
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9356 non-null   float64
dtypes: float64(7), int64(6), object(2)
memory usage: 1.1+ MB


#### I do some data preprocessing like Data Cleaning and Data Reduction.

Data Reduction
The dataset have 15 columns and 9357 rows. I choose 12 columns for data mining. I'm not using date, time, and C6H6(GT) attributes. You can see description about the attributes, and although maybe this dataset for timeseries, but i'm not using date and time. Whereas C6H6(GT) haven't sensor measure result, since it's not relevant with another attributes that have actual and sensor measures. But, this is just my opinion.
Data Cleaning
There are some missing values in this dataset, there are 200 and -200 values in each column, that's missing value. This missing value can make a bad classify result because the value sometimes not relevant with range in some columns. I fill the missing value with mean or median value from each columns. But, first i check if it's a null value.

In [5]:
#reduce columns
df = df.drop('Date', axis=1)
df = df.drop('Time', axis=1)
df = df.drop('C6H6(GT)', axis=1)

In [6]:
#check null value
df.isnull().sum()

CO(GT)           0
PT08.S1(CO)      0
NMHC(GT)         0
PT08.S2(NMHC)    0
NOx(GT)          0
PT08.S3(NOx)     0
NO2(GT)          0
PT08.S4(NO2)     0
PT08.S5(O3)      0
T                0
RH               0
AH               1
dtype: int64

In [7]:
#check minimum values each attribute
df.describe()

CO(GT)  PT08.S1(CO)  ...           RH           AH
count  9357.000000  9357.000000  ...  9357.000000  9356.000000
mean    -34.033536  1048.816074  ...    39.479397    -6.838464
std      79.096080   330.062067  ...    51.217345    38.978664
min    -200.000000  -200.000000  ...  -200.000000  -200.000000
25%       0.600000   921.000000  ...    34.100000     0.692250
50%       1.500000  1053.000000  ...    48.500000     0.976800
75%       2.600000  1221.000000  ...    61.900000     1.296300
max    1428.000000  2040.000000  ...    88.700000     2.231000

[8 rows x 12 columns]

In [ ]:
for i in df.ix[df['CO(GT)']<0].index:
    df.loc[i, 'CO(GT)'] = 0
for i in df.ix[df['PT08.S1(CO)']<0].index:
    df.loc[i, 'PT08.S1(CO)'] = 0
for i in df.ix[df['NMHC(GT)']<0].index:
    df.loc[i, 'NMHC(GT)'] = 0
for i in df.ix[df['PT08.S2(NMHC)']<0].index:
    df.loc[i, 'PT08.S2(NMHC)'] = 0
for i in df.ix[df['NOx(GT)']<0].index:
    df.loc[i, 'NOx(GT)'] = 0
for i in df.ix[df['PT08.S3(NOx)']<0].index:
    df.loc[i, 'PT08.S3(NOx)'] = 0
for i in df.ix[df['NO2(GT)']<0].index:
    df.loc[i, 'NO2(GT)'] = 0
for i in df.ix[df['PT08.S4(NO2)']<0].index:
    df.loc[i, 'PT08.S4(NO2)'] = 0
for i in df.ix[df['PT08.S5(O3)']<0].index:
    df.loc[i, 'PT08.S5(O3)'] = 0
for i in df.ix[df['RH']<0].index:
    df.loc[i, 'RH'] = 0
for i in df.ix[df['AH']<0].index:
    df.loc[i, 'AH'] = 0
#this case because it's possible the Temperature have minus degree,
#so i just fill 0 that have -200 value
for i in df.ix[df['T']==-200].index:
    df.loc[i, 'T'] = 0

In [9]:
#if you check min value has change
df['CO(GT)'].min()

-200.0

In [10]:
#mean values has change
df.describe()

CO(GT)  PT08.S1(CO)  ...           RH           AH
count  9357.000000  9357.000000  ...  9357.000000  9356.000000
mean    -34.033536  1048.816074  ...    39.479397    -6.838464
std      79.096080   330.062067  ...    51.217345    38.978664
min    -200.000000  -200.000000  ...  -200.000000  -200.000000
25%       0.600000   921.000000  ...    34.100000     0.692250
50%       1.500000  1053.000000  ...    48.500000     0.976800
75%       2.600000  1221.000000  ...    61.900000     1.296300
max    1428.000000  2040.000000  ...    88.700000     2.231000

[8 rows x 12 columns]

In [11]:
df['CO(GT)'].median()

1.5

In [12]:
df['PT08.S1(CO)'].median()

1053.0

In [13]:
df['NMHC(GT)'].median()

-200.0

In [14]:
df['PT08.S2(NMHC)'].median()

895.0

In [15]:
df['NOx(GT)'].median()

141.0

In [16]:
df['PT08.S3(NOx)'].median()

794.0

In [17]:
df['NO2(GT)'].median()

96.0

In [18]:
df['PT08.S4(NO2)'].median()

1446.0

In [19]:
df['PT08.S5(O3)'].median()

942.0

In [20]:
df['T'].median()


17.2

In [21]:
df['RH'].median()


48.5

In [22]:
df['AH'].median()


0.9768

In [ ]:
for i in df.ix[df['CO(GT)']==0].index:
    df.loc[i, 'CO(GT)'] = df['CO(GT)'].mean()
for i in df.ix[df['PT08.S1(CO)']==0].index:
    df.loc[i, 'PT08.S1(CO)'] = df['PT08.S1(CO)'].mean()
for i in df.ix[df['NMHC(GT)']==0].index:
    df.loc[i, 'NMHC(GT)'] = df['NMHC(GT)'].mean()
for i in df.ix[df['PT08.S2(NMHC)']==0].index:
    df.loc[i, 'PT08.S2(NMHC)'] = df['PT08.S2(NMHC)'].mean()
for i in df.ix[df['NOx(GT)']==0].index:
    df.loc[i, 'NOx(GT)'] = df['NOx(GT)'].median()
for i in df.ix[df['PT08.S3(NOx)']==0].index:
    df.loc[i, 'PT08.S3(NOx)'] = df['PT08.S3(NOx)'].mean()
for i in df.ix[df['NO2(GT)']==0].index:
    df.loc[i, 'NO2(GT)'] = df['NO2(GT)'].mean()
for i in df.ix[df['PT08.S4(NO2)']==0].index:
    df.loc[i, 'PT08.S4(NO2)'] = df['PT08.S4(NO2)'].median()
for i in df.ix[df['PT08.S5(O3)']==0].index:
    df.loc[i, 'PT08.S5(O3)'] = df['PT08.S5(O3)'].median()
for i in df.ix[df['T']==0].index:
    df.loc[i, 'T'] = df['T'].mean()
for i in df.ix[df['RH']==0].index:
    df.loc[i, 'RH'] = df['RH'].mean()
for i in df.ix[df['AH']==0].index:
    df.loc[i, 'AH'] = df['AH'].mean()

In [24]:
df['CO(GT)'].min()


-200.0

In [25]:
print(df[df['AH'].isnull()])


      CO(GT)  PT08.S1(CO)  NMHC(GT)  ...     T      RH  AH
4887  1428.0         -200      20.9  ...  57.2  1.2146 NaN

[1 rows x 12 columns]


In [26]:
df['AH'].iloc[4887] = df['AH'].mean()


/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [27]:
from sklearn.cluster import KMeans

km = KMeans(n_clusters=2, random_state=1)
new = df._get_numeric_data().dropna(axis=1)
km.fit(new)
predict=km.predict(new)

In [28]:
df['Class'] = pd.Series(predict, index=df.index)


In [29]:
df.head()


CO(GT)  PT08.S1(CO)  NMHC(GT)  PT08.S2(NMHC)  ...     T    RH      AH  Class
0     2.6         1360     150.0           1046  ...  13.6  48.9  0.7578      1
1     2.0         1292     112.0            955  ...  13.3  47.7  0.7255      1
2     2.2         1402      88.0            939  ...  11.9  54.0  0.7502      1
3     2.2         1376      80.0            948  ...  11.0  60.0  0.7867      1
4     1.6         1272      51.0            836  ...  11.2  59.6  0.7888      1

[5 rows x 13 columns]

In [30]:
X = df.drop('Class', axis=1)
y = df.loc[:,'Class'].values

In [33]:
from sklearn.model_selection import train_test_split
#split data with 70%, 30%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,  random_state =0)

In [34]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [36]:
#use nn for classify
from sklearn.linear_model import Perceptron


In [ ]:
ppn =  Perceptron(n_init=100, eta0=0.1, random_state=0)
ppn.fit(X_train_std, y_train)

In [ ]:
y_pred = ppn.predict(X_test_std)
print('Misclassified sample : %d' % (y_test != y_pred).sum())

from sklearn.metrics import accuracy_score
print('accuracy : %.2f' % accuracy_score(y_test, y_pred))



Misclassified sample : 10
accuracy : 1.00